beacon_node/network/src/attestation_service/mod.rs

```diff
@@ -186,18 +186,49 @@ impl<T: BeaconChainTypes> AttestationService<T> {
     pub fn validator_subscriptions(
         &mut self,
         subscriptions: Vec<ValidatorSubscription>,
-    ) -> Result<(), ()> {
+    ) -> Result<(), String> {
+        let current_slot = self
+            .beacon_chain
+            .slot()
+            .map_err(|e| format!("Failed to get current slot: {:?}", e))?;
+
+        let mut state = self
+            .beacon_chain
+            .state_at_slot(current_slot, StateSkipConfig::WithoutStateRoots)
+            .map_err(|e| format!("Failed to get beacon state: {:?}", e))?;
+
+        // Ensure that the committee caches are built
+        state
```

Comment:
@michaelsproul I think we just need the `CommitteeCache` for `RelativeEpoch::Next` here and below.
Do we need to build all the committee caches here?

Reply:
I'm not super familiar with the attestation service so let me just see if I understand the assumptions in place here:

1. `validator_subscriptions` will be called upon prompting from the validator client (via the REST API). We assume it notifies us in the epoch prior to the one in which we must attest.
2. The slot clock of the beacon node matches the validator client, so the state at the beacon chain's current slot will be one from the epoch prior to the attestation epoch.
3. All the `ValidatorSubscription`s have slots in the same epoch.

I think these assumptions are likely to hold "most of the time". In that case, we could avoid some extra processing by using the `BeaconChain`'s head state instead of a `state_at_slot` state (which is slower to construct, and doesn't come with caches pre-built -- the head state does). If the current slot is the same epoch as the head, which should be the case most of the time, we could do something like:

```rust
let head_epoch = self.beacon_chain.head_info()?.slot.epoch(..);
let current_slot = self.beacon_chain.slot()?;

let state = if current_slot.epoch(..) == head_epoch {
    self.beacon_chain.head()?.beacon_state
} else {
    let mut state = self.beacon_chain.state_at_slot(current_slot, StateSkipConfig::WithoutStateRoots)?;
    state.build_committee_cache(RelativeEpoch::Next, &self.beacon_chain.spec)?;
    state
};
```

It's a bit more mucking around, but it saves building the committee cache or processing unnecessary slots in the most common case. It _might_ be a premature optimization though, so maybe we could try just building the cache and see how that goes. Either way we're cloning a whole `BeaconState`, which might impact memory usage.

Would be interested to hear @paulhauner's thoughts on this if he has time, as he's done a lot of work to avoid cloning states in the past.

Reply:
As discussed out-of-band with @michaelsproul, there's a race condition in the above code snippet since there's no guarantee that the head won't change between calls to `self.beacon_chain.head_info()` and `self.beacon_chain.head()`.